In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.utils import plot_model
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [2]:
print("TensorFlow v" + tf.__version__)

TensorFlow v2.12.0


In [3]:
data = pd.read_csv('starcraft_player_data.csv')
print("Full train dataset shape is {}".format(data.shape))

Full train dataset shape is (3395, 20)


#### The data is composed of 20 columns and 3395 entries. We can see all 20 dimensions of the dataset by printing the first 5 entries using the following code:

In [4]:
# display first 5 examples
data.head(5)

GameID  LeagueIndex Age HoursPerWeek TotalHours       APM  SelectByHotkeys  \
0      52            5  27           10       3000  143.7180         0.003515   
1      55            5  23           10       5000  129.2322         0.003304   
2      56            4  30           10        200   69.9612         0.001101   
3      57            3  19           20        400  107.6016         0.001034   
4      58            3  32           10        500  122.8908         0.001136   

   AssignToHotkeys  UniqueHotkeys  MinimapAttacks  MinimapRightClicks  \
0         0.000220              7        0.000110            0.000392   
1         0.000259              4        0.000294            0.000432   
2         0.000336              4        0.000294            0.000461   
3         0.000213              1        0.000053            0.000543   
4         0.000327              2        0.000000            0.001329   

   NumberOfPACs  GapBetweenPACs  ActionLatency  ActionsInPAC  \
0      0.004849         32.6677        40.8673        4.7508   
1      0.004307         32.9194        42.3454        4.8434   
2      0.002926         44.6475        75.3548        4.0430   
3      0.003783         29.2203        53.7352        4.9155   
4      0.002368         22.6885        62.0813        9.3740   

   TotalMapExplored  WorkersMade  UniqueUnitsMade  ComplexUnitsMade  \
0                28     0.001397                6               0.0   
1                22     0.001194                5               0.0   
2                22     0.000745                6               0.0   
3                19     0.000426                7               0.0   
4                15     0.001174                4               0.0   

   ComplexAbilitiesUsed  
0              0.000000  
1              0.000208  
2              0.000189  
3              0.000384  
4              0.000019

In [5]:
data.info()
data.describe(include='all').T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3395 entries, 0 to 3394
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   GameID                3395 non-null   int64  
 1   LeagueIndex           3395 non-null   int64  
 2   Age                   3395 non-null   object 
 3   HoursPerWeek          3395 non-null   object 
 4   TotalHours            3395 non-null   object 
 5   APM                   3395 non-null   float64
 6   SelectByHotkeys       3395 non-null   float64
 7   AssignToHotkeys       3395 non-null   float64
 8   UniqueHotkeys         3395 non-null   int64  
 9   MinimapAttacks        3395 non-null   float64
 10  MinimapRightClicks    3395 non-null   float64
 11  NumberOfPACs          3395 non-null   float64
 12  GapBetweenPACs        3395 non-null   float64
 13  ActionLatency         3395 non-null   float64
 14  ActionsInPAC          3395 non-null   float64
 15  TotalMapExplored     

count unique  top freq         mean          std  \
GameID                3395    NaN  NaN  NaN      4805.01      2719.94   
LeagueIndex           3395    NaN  NaN  NaN      4.18409      1.51733   
Age                   3395     29   20  357          NaN          NaN   
HoursPerWeek          3395     33   10  411          NaN          NaN   
TotalHours            3395    238  500  328          NaN          NaN   
APM                   3395    NaN  NaN  NaN      117.047      51.9453   
SelectByHotkeys       3395    NaN  NaN  NaN   0.00429866   0.00528443   
AssignToHotkeys       3395    NaN  NaN  NaN  0.000373576  0.000224934   
UniqueHotkeys         3395    NaN  NaN  NaN      4.36465      2.36033   
MinimapAttacks        3395    NaN  NaN  NaN  9.83067e-05  0.000165871   
MinimapRightClicks    3395    NaN  NaN  NaN  0.000387436  0.000377001   
NumberOfPACs          3395    NaN  NaN  NaN   0.00346289  0.000992293   
GapBetweenPACs        3395    NaN  NaN  NaN      40.3616      17.1536   
ActionLatency         3395    NaN  NaN  NaN      63.7394      19.2389   
ActionsInPAC          3395    NaN  NaN  NaN      5.27299      1.49484   
TotalMapExplored      3395    NaN  NaN  NaN      22.1317      7.43172   
WorkersMade           3395    NaN  NaN  NaN   0.00103169  0.000519008   
UniqueUnitsMade       3395    NaN  NaN  NaN      6.53402       1.8577   
ComplexUnitsMade      3395    NaN  NaN  NaN  5.94327e-05  0.000111091   
ComplexAbilitiesUsed  3395    NaN  NaN  NaN  0.000141878  0.000265247   

                              min          25%          50%          75%  \
GameID                         52       2464.5         4874       7108.5   
LeagueIndex                     1            3            4            5   
Age                           NaN          NaN          NaN          NaN   
HoursPerWeek                  NaN          NaN          NaN          NaN   
TotalHours                    NaN          NaN          NaN          NaN   
APM                       22.0596      79.9002       108.01       142.79   
SelectByHotkeys                 0   0.00125782   0.00249951   0.00513257   
AssignToHotkeys                 0  0.000204226  0.000352568  0.000498834   
UniqueHotkeys                   0            3            4            6   
MinimapAttacks                  0            0  3.99297e-05  0.000118854   
MinimapRightClicks              0  0.000140086  0.000281489  0.000514095   
NumberOfPACs          0.000678996    0.0027537   0.00339542    0.0040271   
GapBetweenPACs             6.6667      28.9577      36.7235      48.2905   
ActionLatency             24.0936      50.4466      60.9318      73.6813   
ActionsInPAC               2.0389      4.27285       5.0955       6.0336   
TotalMapExplored                5           17           22           27   
WorkersMade            7.6982e-05   0.00068296   0.00090524   0.00125865   
UniqueUnitsMade                 2            5            6            8   
ComplexUnitsMade                0            0            0  8.55403e-05   
ComplexAbilitiesUsed            0            0   2.0282e-05   0.00018137   

                              max  
GameID                      10095  
LeagueIndex                     8  
Age                           NaN  
HoursPerWeek                  NaN  
TotalHours                    NaN  
APM                       389.831  
SelectByHotkeys         0.0430884  
AssignToHotkeys        0.00175219  
UniqueHotkeys                  10  
MinimapAttacks         0.00301935  
MinimapRightClicks      0.0040408  
NumberOfPACs           0.00797064  
GapBetweenPACs            237.143  
ActionLatency             176.372  
ActionsInPAC              18.5581  
TotalMapExplored               58  
WorkersMade             0.0051493  
UniqueUnitsMade                13  
ComplexUnitsMade      0.000902274  
ComplexAbilitiesUsed    0.0030837

In [6]:
# separate numerical and categorical variables for easy analysis
cat_cols = data.select_dtypes(include=['object']).columns
num_cols = data.select_dtypes(include=np.number).columns.tolist()

print("Categorical values: ")
print(cat_cols)
print("Numerical values: ")
print(num_cols)

Categorical values: 
Index(['Age', 'HoursPerWeek', 'TotalHours'], dtype='object')
Numerical values: 
['GameID', 'LeagueIndex', 'APM', 'SelectByHotkeys', 'AssignToHotkeys', 'UniqueHotkeys', 'MinimapAttacks', 'MinimapRightClicks', 'NumberOfPACs', 'GapBetweenPACs', 'ActionLatency', 'ActionsInPAC', 'TotalMapExplored', 'WorkersMade', 'UniqueUnitsMade', 'ComplexUnitsMade', 'ComplexAbilitiesUsed']


In [7]:
# default performance = 0
performance_values = 0

# add the 'performance' column to data
data['PlayerPerformance'] = performance_values
data.to_csv('your_dataset.csv', index=False)

data.head()

GameID  LeagueIndex Age HoursPerWeek TotalHours       APM  SelectByHotkeys  \
0      52            5  27           10       3000  143.7180         0.003515   
1      55            5  23           10       5000  129.2322         0.003304   
2      56            4  30           10        200   69.9612         0.001101   
3      57            3  19           20        400  107.6016         0.001034   
4      58            3  32           10        500  122.8908         0.001136   

   AssignToHotkeys  UniqueHotkeys  MinimapAttacks  ...  NumberOfPACs  \
0         0.000220              7        0.000110  ...      0.004849   
1         0.000259              4        0.000294  ...      0.004307   
2         0.000336              4        0.000294  ...      0.002926   
3         0.000213              1        0.000053  ...      0.003783   
4         0.000327              2        0.000000  ...      0.002368   

   GapBetweenPACs  ActionLatency  ActionsInPAC  TotalMapExplored  WorkersMade  \
0         32.6677        40.8673        4.7508                28     0.001397   
1         32.9194        42.3454        4.8434                22     0.001194   
2         44.6475        75.3548        4.0430                22     0.000745   
3         29.2203        53.7352        4.9155                19     0.000426   
4         22.6885        62.0813        9.3740                15     0.001174   

   UniqueUnitsMade  ComplexUnitsMade  ComplexAbilitiesUsed  PlayerPerformance  
0                6               0.0              0.000000                  0  
1                5               0.0              0.000208                  0  
2                6               0.0              0.000189                  0  
3                7               0.0              0.000384                  0  
4                4               0.0              0.000019                  0  

[5 rows x 21 columns]

In [8]:
# separate features and target variable
categorical_cols = ['Age', 'HoursPerWeek', 'TotalHours']
numerical_cols = ['GameID', 'LeagueIndex', 'APM', 'SelectByHotkeys', 'AssignToHotkeys', 'UniqueHotkeys', 'MinimapAttacks', 'MinimapRightClicks', 'NumberOfPACs', 'GapBetweenPACs', 'ActionLatency', 'ActionsInPAC', 'TotalMapExplored', 'WorkersMade', 'UniqueUnitsMade', 'ComplexUnitsMade', 'ComplexAbilitiesUsed']
target_col = 'PlayerPerformance'

# convert categorical variables to encoding
data = pd.get_dummies(data, columns=categorical_cols)

# scale numerical variables
scaler = StandardScaler()
data[numerical_cols] = scaler.fit_transform(data[numerical_cols])

# split the data into training and testing sets
x = data.drop(target_col, axis=1).values
y = data[target_col].values

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [9]:
model = tf.keras.Sequential([
    tf.keras.layers.LSTM(64, input_shape=(x_train.shape[1], 1)),
    tf.keras.layers.Dense(1)
])

model.compile(loss='mean_squared_error', optimizer='adam')

In [10]:
x_train_rnn = np.expand_dims(x_train, axis=2)  # reshape input

model.fit(x_train_rnn, y_train, epochs=10, batch_size=32)

Epoch 1/10
85/85 [==============================] - 8s 75ms/step - loss: 6.4361e-06
Epoch 2/10
85/85 [==============================] - 7s 78ms/step - loss: 1.9624e-09
Epoch 3/10
85/85 [==============================] - 7s 78ms/step - loss: 6.0305e-10
Epoch 4/10
85/85 [==============================] - 7s 78ms/step - loss: 6.2893e-10
Epoch 5/10
85/85 [==============================] - 7s 79ms/step - loss: 3.7595e-10
Epoch 6/10
85/85 [==============================] - 7s 78ms/step - loss: 5.1682e-10
Epoch 7/10
85/85 [==============================] - 7s 81ms/step - loss: 5.2198e-10
Epoch 8/10
85/85 [==============================] - 9s 106ms/step - loss: 3.3949e-10
Epoch 9/10
85/85 [==============================] - 8s 91ms/step - loss: 2.6075e-10
Epoch 10/10
85/85 [==============================] - 7s 84ms/step - loss: 6.6198e-10


In [11]:
x_test_rnn = np.expand_dims(x_test, axis=2)  # reshape input for RNN

predictions = model.predict(x_test_rnn)
mse = np.mean((predictions - y_test) ** 2)
print("Mean Squared Error:", mse)

22/22 [==============================] - 1s 24ms/step
Mean Squared Error: 5.463003204584331e-10


In [ ]:
# define RNN model
model = tf.keras.Sequential([
    tf.keras.layers.LSTM(64, input_shape=(x_train.shape[1], 1)),
    tf.keras.layers.Dense(1)
])

# plot the model
plot_model(model, to_file='model_rnn.png', show_shapes=True)

#### Hypothetical: the model would be better if we had the rank of the players from previous years / competitions. The first test would train the neural map using the previous performance of the player and then predict their 2023 ranking, looking only one game into the future at a time, meaning the map would benefit from all real data up to the game in question. The graph would then show the predictions compared to the real data. The two graph lines shouldn't match up exactly because that would mean the RNN memorized the training data rather than create a prediction. 